services/mpa/server/server.go

```diff
@@ -0,0 +1,291 @@
+		if !ok {
+			// Nothing to look up if MPA wasn't requested
+			return nil
+		}
+		resp, err := serverSingleton.Get(ctx, &mpa.GetRequest{Id: mpaID})
+		if err != nil {
+			return err
+		}
+
+		if err := mpahooks.ActionMatchesInput(resp.Action, input); err != nil {
+			return err
+		}
+		for _, a := range resp.Approver {
+			input.Approvers = append(input.Approvers, &rpcauth.PrincipalAuthInput{
+				ID:     a.Id,
+				Groups: a.Groups,
+			})
+		}
+		return nil
+	})
+}
+
+// actionId generates the id for an action by hashing it
+func actionId(action *mpa.Action) (string, error) {
+	// Binary proto encoding doesn't provide any guarantees about deterministic
+	// output for the same input. Go provides a deterministic marshalling option,
+	// but this marshalling isn't guaranteed to be stable over time.
+	// JSON encoding can be made deterministic by canonicalizing.
+	b, err := protojson.Marshal(action)
+	if err != nil {
+		return "", err
+	}
+	canonical, err := jcs.Transform(b)
+	if err != nil {
+		return "", err
+	}
+	h := sha256.New()
+	h.Write(canonical)
+	sum := h.Sum(nil)
+	// Humans are going to need to deal with these, so let's shorten them
+	// and make them a bit prettier.
+	return fmt.Sprintf("%x-%x-%x", sum[0:4], sum[4:8], sum[8:12]), nil
+
+}
+
+type storedAction struct {
+	action       *mpa.Action
+	lastModified time.Time
+	approvers    []*mpa.Principal
+	approved     chan struct{}
+}
+
+// server is used to implement the gRPC server
+type server struct {
+	actions map[string]*storedAction
+	mu      sync.Mutex
+}
+
+func callerIdentity(ctx context.Context) (*rpcauth.PrincipalAuthInput, bool) {
+	// TODO(#346): Prefer using a proxied identity if provided
+	peer := rpcauth.PeerInputFromContext(ctx)
+	if peer != nil {
+		return peer.Principal, true
+	}
+	return nil, false
+}
+
+func (s *server) clearOutdatedApprovals() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	staleTime := time.Now().Add(-maxMPAApprovedAge)
+	for id, act := range s.actions {
+		if act.lastModified.Before(staleTime) {
+			delete(s.actions, id)
+		}
+	}
+
+}
+
+func (s *server) Store(ctx context.Context, in *mpa.StoreRequest) (*mpa.StoreResponse, error) {
+	var justification string
+	if md, found := metadata.FromIncomingContext(ctx); found && len(md[rpcauth.ReqJustKey]) > 0 {
+		justification = md[rpcauth.ReqJustKey][0]
+	}
+
+	p, ok := callerIdentity(ctx)
+	if !ok || p == nil {
+		return nil, status.Error(codes.FailedPrecondition, "unable to determine caller's identity")
+	}
+
+	action := &mpa.Action{
+		User:          p.ID,
+		Justification: justification,
+		Method:        in.Method,
+		Message:       in.Message,
+	}
+	id, err := actionId(action)
+	if err != nil {
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Time to clear out excessive approvals!
+	for len(s.actions) >= maxMPAApprovals {
+		var oldestID string
+		oldestTime := time.Now()
+		for id, act := range s.actions {
+			if act.lastModified.Before((oldestTime)) {
```

Comment:
nit: what's with the extra parens around 'oldestTime'?

Reply:
Huh, that's a typo that happens to compile. Fixed.